src/linalg.jl

```diff
@@ -220,108 +166,96 @@ end
     return quote
         $(Expr(:meta, :inline))
-        $(Expr(:call, SM, Expr(:tuple, e...)))
+        SM(tuple($(exprs...)))
     end
 end
 
-@generated function diagm(v::StaticVector)
+@inline diagm(v::StaticVector) = _diagm(Size(v), v)
+@generated function _diagm(::Size{S}, v::StaticVector) where {S}
+    Snew = (S[1], S[1])
     T = eltype(v)
-    exprs = [i == j ? :(v[$i]) : zero(T) for i = 1:length(v), j = 1:length(v)]
-    newtype = similar_type(v, Size(length(v), length(v)))
+    exprs = [i == j ? :(v[$i]) : zero(T) for i = 1:S[1], j = 1:S[1]]
     return quote
         $(Expr(:meta, :inline))
-        @inbounds return $(Expr(:call, newtype, Expr(:tuple, exprs...)))
+        @inbounds return similar_type($v, Size($Snew))(tuple($(exprs...)))
     end
 end
 
-@generated function cross(a::StaticVector, b::StaticVector)
-    S = typeof(zero(eltype(a))*zero(eltype(b)))
-    if length(a) === 3 && length(b) === 3
-        if S <: Unsigned
-            return quote
-                $(Expr(:meta, :inline))
-                similar_type(a, $(typeof(Signed(zero(eltype(a))*zero(eltype(b))))))((Signed(a[2]*b[3])-Signed(a[3]*b[2]), Signed(a[3]*b[1])-Signed(a[1]*b[3]), Signed(a[1]*b[2])-Signed(a[2]*b[1])))
-            end
-        else
-            return quote
-                $(Expr(:meta, :inline))
-                similar_type(a, $S)((a[2]*b[3]-a[3]*b[2], a[3]*b[1]-a[1]*b[3], a[1]*b[2]-a[2]*b[1]))
-            end
-        end
-    elseif length(a) === 2 && length(b) === 2
-        return quote
-            $(Expr(:meta, :inline))
-            promote_type(eltype(a), eltype(b))(a[1]*b[2]-a[2]*b[1])
-        end
-    elseif length(a) === 2 && length(b) === 2
-        return quote
-            $(Expr(:meta, :inline))
-            promote_type(eltype(a), eltype(b))(a[1]*b[2]-a[2]*b[1])
-        end
-    else
-        error("Cross product only defined for 3-vectors")
-    end
+@inline cross(a::StaticVector, b::StaticVector) = _cross(same_size(a, b), a, b)
+_cross(::Size{S}, a::StaticVector, b::StaticVector) where {S} = error("Cross product not defined for $(S[1])-vectors")
+@inline function _cross(::Size{(2,)}, a::StaticVector, b::StaticVector)
+    @inbounds return a[1]*b[2] - a[2]*b[1]
+end
+@inline function _cross(::Size{(3,)}, a::StaticVector, b::StaticVector)
+    T = typeof(a[2]*b[3]-a[3]*b[2])
+    @inbounds return similar_type(a, typeof(a[2]*b[3]-a[3]*b[2]))((a[2]*b[3]-a[3]*b[2], a[3]*b[1]-a[1]*b[3], a[1]*b[2]-a[2]*b[1]))
+end
+@inline function _cross(::Size{(2,)}, a::StaticVector{<:Unsigned}, b::StaticVector{<:Unsigned})
+    @inbounds return Signed(a[1]*b[2]) - Signed(a[2]*b[1])
+end
+@inline function _cross(::Size{(3,)}, a::StaticVector{<:Unsigned}, b::StaticVector{<:Unsigned})
+    T = typeof(a[2]*b[3]-a[3]*b[2])
+    @inbounds return similar_type(a, typeof(Signed(a[2]*b[3])-Signed(a[3]*b[2])))(((Signed(a[2]*b[3])-Signed(a[3]*b[2]), Signed(a[3]*b[1])-Signed(a[1]*b[3]), Signed(a[1]*b[2])-Signed(a[2]*b[1]))))
```

Comment:
Yeah, it just happens to be needed for the determinant algorithm, and someone raised an issue about that! Seemed a bit silly but oh well.